### PR TITLE
Issue #1952 - OperationOutcome when search parameters are ignored

### PR DESCRIPTION
**Pull request by tbieste**
Issue #1952 - OperationOutcome when search parameters are ignored

Signed-off-by: Troy Biesterfeld <tbieste@us.ibm.com>

### CODE REVIEW

File: fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java [hunk 1]
```diff
@@ -81,72 +85,77 @@ public static Direction fromValue(String value) {
...
 
-                checkSystemLevel(resourceTypeName, sortParm.getCode());
+                    checkSystemLevel(resourceTypeName, sortParm.getCode(), context);
 
-                context.getSortParameters().add(sortParm);
+                    context.getSortParameters().add(sortParm);
+                }
+            } catch (FHIRSearchException se) {
+                // If we're in lenient mode and there was an issue parsing the _sort parameter value then log and move on to the next one.
+                if (context.isLenient()) {
```
Comment (prb112) [inline fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java:128]: Where is this set?

File: fhir-search/src/test/java/com/ibm/fhir/search/location/NearLocationHandlerBoundingRadiusTest.java [hunk 2]
```diff
@@ -21,12 +21,13 @@
 import com.ibm.fhir.search.location.bounding.Bounding;
 import com.ibm.fhir.search.location.bounding.BoundingRadius;
 import com.ibm.fhir.search.location.bounding.BoundingType;
+import com.ibm.fhir.search.test.BaseSearchTest;
 import com.ibm.fhir.search.util.SearchUtil;
 
 /**
  * Test the BoundingRadius with NearLocationHandler
  */
-public class NearLocationHandlerBoundingRadiusTest {
```
Comment (prb112) [inline fhir-search/src/test/java/com/ibm/fhir/search/location/NearLocationHandlerBoundingRadiusTest.java:29]: Why the extends?

File: fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java (hunk 1 above)
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java:128]: It is set at the top of SearchUtil.parseQueryParameters(...), which calls SearchUtil.parseSearchResultParameter(...), which calls this method.

File: fhir-search/src/test/java/com/ibm/fhir/search/location/NearLocationHandlerBoundingRadiusTest.java (hunk 2 above)
Comment (tbieste) [reply fhir-search/src/test/java/com/ibm/fhir/search/location/NearLocationHandlerBoundingRadiusTest.java:29]: Depending on the order the tests are run, sometimes FHIRRequestContext was null, which causes a NPE down within the call to parseQueryParameters(...), so I extended BaseSearchTest to ensure it was set.

File: fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java [hunk 3]
```diff
@@ -81,72 +85,77 @@ public static Direction fromValue(String value) {
...
+            FHIRSearchContext context) throws FHIRSearchException {
         boolean result = sortParmProxy == null;
         if (result) {
             String msg = buildUndefinedSortParamMessage(resourceTypeName, sortParmCode);
-
-            // Stop Processing if not lenient
-            if (!lenient) {
-                throw SearchExceptionUtil.buildNewInvalidSearchException(msg);
+            if (context.isLenient()) {
+                log.fine(msg);
```
Comment (prb112) [inline fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java:155]: isLoggable?

File: fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java [hunk 4]
```diff
@@ -81,72 +85,77 @@ public static Direction fromValue(String value) {
...
 
-                context.getSortParameters().add(sortParm);
+                    context.getSortParameters().add(sortParm);
+                }
+            } catch (FHIRSearchException se) {
+                // If we're in lenient mode and there was an issue parsing the _sort parameter value then log and move on to the next one.
+                if (context.isLenient()) {
+                    String msg = "Sort value '" + sortParmCode + "' for resource type '" + resourceTypeName + "' ignored";
+                    log.log(Level.FINE, msg, se);
+                    context.addOutcomeIssue(FHIRUtil.buildOperationOutcomeIssue(IssueSeverity.WARNING, IssueType.INCOMPLETE, msg));
```
Comment (prb112) [inline fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java:131]: Previously we threw an error?  this is pretty aggressive for lenient... do we want this behavior change?

File: fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java [hunk 5]
```diff
@@ -157,13 +166,18 @@ public boolean isUndefinedOrLenient(String resourceTypeName, String sortParmCode
...
      */
-    public void checkSystemLevel(String resourceTypeName, String code) throws FHIRSearchException {
+    public void checkSystemLevel(String resourceTypeName, String code, FHIRSearchContext context) throws FHIRSearchException {
         if ("Resource".compareTo(resourceTypeName) == 0
                 && !SYSTEM_LEVEL_SORT_PARAMETER_NAMES.contains(code)) {
-            throw SearchExceptionUtil.buildNewInvalidSearchException(
-                    buildNewInvalidSearchExceptionMessage(code));
+            String msg = buildNewInvalidSearchExceptionMessage(code);
+            if (context.isLenient()) {
+                log.fine(msg);
```
Comment (prb112) [inline fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java:177]: isLoggable

**Review by prb112 — APPROVED** [review]
A couple comments, looks good

File: fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java (hunk 3 above)
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java:155]: The log.fine(...) method internally does an isLoggable check.

File: fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java (hunk 5 above)
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java:177]: The log.fine(...) method internally does an isLoggable check.

Comment (prb112) [reply fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java:177]: that's correct, as a general pattern we've followed isloggable, more specifically is the msg is annotated or constructed.
it'd be decent to be defensive and not build the msg unless we need to, and log only as needed.

File: fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java (hunk 4 above)
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java:131]: This method is only called by parseSortParameter(...), which in a loop catches the exception, and then moves on to the next sort code. I've updated the JavaDoc for the methods to explain the behavior.

File: fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java (hunk 5 above)
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java:174]: Yeah, I agree.  In this case, the msg always needs to be constructed to be added to the OutcomeIssues and/or the thrown exception, so I didn't add the isLoggable check.

Comment (prb112) [reply fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java:174]: makes sense

File: fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java (hunk 4 above)
Comment (prb112) [reply fhir-search/src/main/java/com/ibm/fhir/search/sort/Sort.java:130]: Seems fine... just worthly of the update to explain, thanks

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 6]
```diff
@@ -891,12 +905,9 @@ public static FHIRSearchContext parseQueryParameters(Class<?> resourceType,
                 // There's a number of places that throw within this try block. In all cases we want the same behavior:
                 // If we're in lenient mode and there was an issue parsing the query parameter then log and move on to the next one.
                 if (lenient) {
-                    // TODO add this to the list of supplemental warnings?
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:894]: love seeing these TODOs get knocked out

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 7]
```diff
@@ -1660,12 +1671,12 @@ private static void parseSearchResultParameter(Class<?> resourceType, FHIRSearch
             } else if (SearchConstants.SORT.equals(name) && first != null) {
                 // in R4, we only look for _sort
                 // Only first value is used, which matches behavior of other parameters that are supposed to be specified at most once
-                sort.parseSortParameter(resourceTypeName, context, first, lenient);
+                sort.parseSortParameter(resourceTypeName, context, first);
             } else if (name.startsWith(SearchConstants.INCLUDE) || name.startsWith(SearchConstants.REVINCLUDE)) {
-                parseInclusionParameter(resourceType, context, name, values, lenient);
+                parseInclusionParameter(resourceType, context, name, values, context.isLenient());
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1676]: any reason not to remove the lenient parameter from parseInclusionParameter as well?

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java [hunk 8]
```diff
@@ -2480,14 +2517,20 @@ public static String buildSearchSelfUri(String requestUriString, FHIRSearchConte
...
+     *     The issue type.
+     * @param context
+     *     The search context.
+     * @param alwaysThrow
+     *     Determine whether exception should be thrown even if in lenient mode.
      * @throws FHIRSearchException
      */
-    private static void manageException(String message, boolean lenient) throws FHIRSearchException {
-        if (lenient) {
+    private static void manageException(String message, IssueType issueType, FHIRSearchContext context, boolean alwaysThrow) throws FHIRSearchException {
```
Comment (lmsurpre) [inline fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:2528]: I don't love the `alwaysThrow` flag, but I guess if we need the method to behave differently in different contexts, then I suppose its need

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java [hunk 9]
```diff
@@ -553,6 +554,51 @@ public void testSearchPatientWithIdentifierOfTypeNotFound() {
...
+        Response response =
+                target.path("Patient").queryParam("unknown", "unknown").request(FHIRMediaType.APPLICATION_FHIR_JSON)
+                .header("Prefer", "handling=lenient")
+                .get();
+        assertResponse(response, Response.Status.OK.getStatusCode());
+        Bundle bundle = response.readEntity(Bundle.class);
+
+        assertNotNull(bundle);
+        assertTrue(bundle.getEntry().size() >= 1);
+        assertTrue(SearchEntryMode.of(bundle.getEntry().get(bundle.getEntry().size()-1).getSearch().getMode().getValue()) == SearchEntryMode.OUTCOME);
```
Comment (lmsurpre) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java:569]: I think either way would work, but I prefer the following:
```suggestion
        assertTrue(bundle.getEntry().get(bundle.getEntry().size()-1).getSearch().getMode().getValueAsEnum()) == SearchEntryMode.ValueSet.OUTCOME);
```

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java [hunk 10]
```diff
@@ -553,6 +554,51 @@ public void testSearchPatientWithIdentifierOfTypeNotFound() {
...
+        Response response =
+                target.path("Patient").queryParam("elements", "unknown").request(FHIRMediaType.APPLICATION_FHIR_JSON)
+                .header("Prefer", "handling=lenient")
+                .get();
+        assertResponse(response, Response.Status.OK.getStatusCode());
+        Bundle bundle = response.readEntity(Bundle.class);
+
+        assertNotNull(bundle);
+        assertTrue(bundle.getEntry().size() >= 1);
+        assertTrue(SearchEntryMode.of(bundle.getEntry().get(bundle.getEntry().size()-1).getSearch().getMode().getValue()) == SearchEntryMode.OUTCOME);
```
Comment (lmsurpre) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java:584]: same

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java [hunk 11]
```diff
@@ -553,6 +554,51 @@ public void testSearchPatientWithIdentifierOfTypeNotFound() {
...
+        Response response =
+                target.path("Patient").queryParam("sort", "unknown").request(FHIRMediaType.APPLICATION_FHIR_JSON)
+                .header("Prefer", "handling=lenient")
+                .get();
+        assertResponse(response, Response.Status.OK.getStatusCode());
+        Bundle bundle = response.readEntity(Bundle.class);
+
+        assertNotNull(bundle);
+        assertTrue(bundle.getEntry().size() >= 1);
+        assertTrue(SearchEntryMode.of(bundle.getEntry().get(bundle.getEntry().size()-1).getSearch().getMode().getValue()) == SearchEntryMode.OUTCOME);
```
Comment (lmsurpre) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java:599]: same

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 7 above)
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:1676]: Fixed.

File: fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java (hunk 8 above)
Comment (tbieste) [reply fhir-search/src/main/java/com/ibm/fhir/search/util/SearchUtil.java:2528]: Yeah, I couldn't think of a better way to keep this commonly used method in place, but still get both types of behavior.

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java (hunk 9 above)
Comment (tbieste) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java:569]: Updated.

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java (hunk 10 above)
Comment (tbieste) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java:584]: Updated.

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java (hunk 11 above)
Comment (tbieste) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/SearchTest.java:599]: Updated.

**Comment by tbieste** [comment]
Comments addressed. Ready to merge.

**Pull request merged**